src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs

```diff
@@ -328,144 +328,580 @@ private void TrySetUnrecoverableError(SyntaxDiagnosticInfo error)
+            /// rest of the literal. <see langword="false"/> if we were not successful and should stop
+            /// processing.</returns>
+            private bool ScanOpenQuote(
+                out InterpolatedStringKind kind,
+                out int startingDollarSignCount,
+                out int startingQuoteCount)
             {
                 // Handles reading the start of the interpolated string literal (up to where the content begins)
-                var start = _lexer.TextWindow.Position;
-                if (_kind is InterpolatedStringKind.Verbatim)
+                var window = _lexer.TextWindow;
+                var start = window.Position;
+
+                if ((window.PeekChar(0), window.PeekChar(1), window.PeekChar(2)) is ('$', '@', '"') or ('@', '$', '"'))
                 {
-                    // skip past @$ or $@
-                    _lexer.TextWindow.AdvanceChar(2);
+                    // $@" or @$"
+                    //
+                    // Note: we do not consider $@""" as the start of raw-string (in error conditions) as that's a legal
+                    // verbatim string beginning already.
+
+                    kind = InterpolatedStringKind.Verbatim;
+                    startingDollarSignCount = 1;
+                    startingQuoteCount = 1;
+                    window.AdvanceChar(3);
+                    return true;
                 }
-                else
+
+                if ((window.PeekChar(0), window.PeekChar(1), window.PeekChar(2), window.PeekChar(3)) is
+                        ('$', '"', not '"', _) or ('$', '"', '"', not '"'))
                 {
-                    Debug.Assert(_lexer.TextWindow.PeekChar() == '$');
-                    _lexer.TextWindow.AdvanceChar();
+                    // $"...
+                    // $""
+                    // not $"""
+                    kind = InterpolatedStringKind.Normal;
+                    startingDollarSignCount = 1;
+                    startingQuoteCount = 1;
+                    window.AdvanceChar(2);
+                    return true;
                 }
 
-                Debug.Assert(_lexer.TextWindow.PeekChar() == '"');
-                _lexer.TextWindow.AdvanceChar();
+                // From this point we have a raw literal of some sort.
+                var prefixAtCount = _lexer.ConsumeAtSignSequence();
+                startingDollarSignCount = _lexer.ConsumeDollarSignSequence();
+                var suffixAtCount = _lexer.ConsumeAtSignSequence();
+                startingQuoteCount = _lexer.ConsumeQuoteSequence();
+
+                var totalAtCount = prefixAtCount + suffixAtCount;
 
-                openQuoteRange = new Range(start, _lexer.TextWindow.Position);
+                // We should only have gotten here if we had at least two characters that made us think we had an interpolated string.
+                Debug.Assert(totalAtCount + startingDollarSignCount + startingQuoteCount >= 2);
```

Comment:
Can we tighten this assertion? I don't think we should count `@` signs in this assertion, as they don't contribute to the chain that gets us here (ie. identifying a raw string). Could we assert `startingDollarSignCount + startingQuoteCount >= 4`?

Reply:
> Can we tighten this assertion? I don't think we should count @ signs in this assertion, as they don't contribute to the chain that gets us here

Actually, they do contribute.  If you have `@@` or `@@"` etc. then we get here :)

Reply:
Then comment above may be confusing: `// From this point we have a raw literal of some sort.`.

Reply:
agreed.  have tweaked that comment.